NullPointerException in DownloadManager in 9.1.0

### Description of the bug:


I'm seeing a crash that looks like this in 9.1.0:
```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@@aspect_rules_js++npm+lock-v101__lodash__file_.._vendored_lodash-4.17.21.tgz' (requested by nodes 'PACKAGE_LOOKUP:@@aspect_rules_js++npm+lock-v101__lodash__file_.._vendored_lodash-4.17.21.tgz//')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:552)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:435)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.compute(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask$InterruptibleTask.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: net.starlark.java.eval.Starlark$UncheckedEvalException: NullPointerException thrown during Starlark evaluation (repository @lock-v101__lodash__file_.._vendored_lodash-4.17.21.tgz)
	at <starlark>.download(<builtin>:0)
	at <starlark>._download_and_extract_archive(/home/adam/.cache/bazel/_bazel_adam/29a747e212fd2293b5aedcea09254708/external/aspect_rules_js+/npm/private/npm_import.bzl:527)
	at <starlark>._npm_import_rule_impl(/home/adam/.cache/bazel/_bazel_adam/29a747e212fd2293b5aedcea09254708/external/aspect_rules_js+/npm/private/npm_import.bzl:604)
Caused by: java.lang.NullPointerException: Cannot invoke "String.isEmpty()" because "path" is null
	at com.google.devtools.build.lib.vfs.PathFragment.createInternal(PathFragment.java:134)
	at com.google.devtools.build.lib.vfs.PathFragment.create(PathFragment.java:126)
	at com.google.devtools.build.lib.bazel.repository.downloader.DownloadManager.getCandidateFileNames(DownloadManager.java:538)
	at com.google.devtools.build.lib.bazel.repository.downloader.DownloadManager.downloadInExecutor(DownloadManager.java:234)
	at com.google.devtools.build.lib.bazel.repository.downloader.DownloadManager.lambda$startDownload$0(DownloadManager.java:143)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPerTaskExecutor$ThreadBoundFuture.run(Unknown Source)
	at java.base/java.lang.VirtualThread.run(Unknown Source)
```

### Which category does this issue belong to?

_No response_

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


To reproduce it you can check out Aspect's rules_js [repo](https://github.com/aspect-build/rules_js/), cd to `e2e/pnpm_lockfiles` and build `//v101:all`. I bisected the problem to this [commit](https://github.com/bazelbuild/bazel/commit/ff8d723c6d4c3300ee023431e09fb2125bf49ca2), though it's not obvious to me how that commit could cause this. The stack trace above is from that commit.

### Which operating system are you running Bazel on?


Ubuntu 25.10

### What is the output of `bazel info release`?


Starting local Bazel server (9.1.0) and connecting to it... release 9.1.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


```text
git@github.com:acozzette/rules_js.git
40ad11c440833301f3491ded68d60b0f6ca063c2
```

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


https://github.com/bazelbuild/bazel/commit/ff8d723c6d4c3300ee023431e09fb2125bf49ca2

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java
@@ -18,6 +18,7 @@ import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.auth.Credentials;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.CharMatcher;
 import com.google.common.base.MoreObjects;
 import com.google.common.base.Strings;
@@ -515,9 +516,7 @@ public class DownloadManager {
     if (!type.isPresent()) {
       return output;
     }
-    String basename =
-        MoreObjects.firstNonNull(
-            Strings.emptyToNull(PathFragment.create(url.getPath()).getBaseName()), "temp");
+    String basename = MoreObjects.firstNonNull(Strings.emptyToNull(getUrlBaseName(url)), "temp");
     if (!type.get().isEmpty()) {
       String suffix = "." + type.get();
       if (!basename.endsWith(suffix)) {
@@ -530,17 +529,35 @@ public class DownloadManager {
   }
 
   /**
-   * Deterimine the list of filenames to look for in the distdirs. Note that an output name may be
-   * specified that is unrelated to the primary URL. This happens, e.g., when the paramter output is
-   * specified in ctx.download.
+   * Determine the list of filenames to look for in the distdirs. Note that an output name may be
+   * specified that is unrelated to the primary URL. This happens, e.g., when the parameter output
+   * is specified in ctx.download.
    */
-  private static ImmutableSet<String> getCandidateFileNames(URI url, Path destination) {
-    String urlBaseName = PathFragment.create(url.getPath()).getBaseName();
+  @VisibleForTesting
+  static ImmutableSet<String> getCandidateFileNames(URI url, Path destination) {
+    String urlBaseName = getUrlBaseName(url);
     if (!Strings.isNullOrEmpty(urlBaseName) && !urlBaseName.equals(destination.getBaseName())) {
       return ImmutableSet.of(urlBaseName, destination.getBaseName());
     } else {
       return ImmutableSet.of(destination.getBaseName());
     }
+  }
+
+  private static String getUrlBaseName(URI url) {
+    String path = url.getPath();
+    if (path == null && url.isOpaque()) {
+      // Match URL#getPath() behavior for opaque file URIs such as file:../archive.tgz.
+      String rawPath = url.getRawSchemeSpecificPart();
+      int queryStart = rawPath.indexOf('?');
+      if (queryStart != -1) {
+        rawPath = rawPath.substring(0, queryStart);
+      }
+      path =
+          rawPath.isEmpty()
+              ? ""
+              : URI.create(url.getScheme() + ":" + rawPath).getSchemeSpecificPart();
+    }
+    return path == null ? "" : PathFragment.create(path).getBaseName();
   }
 
   private static class CacheProgress implements ExtendedEventHandler.FetchProgress {

--- a/src/test/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloaderTest.java
+++ b/src/test/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloaderTest.java
@@ -179,6 +179,121 @@ public class HttpDownloaderTest {
   }
 
   @Test
+  public void downloadFromOpaqueFileUrl_withExplicitOutput() throws Exception {
+    Downloader downloader = mock(Downloader.class);
+    DownloadManager downloadManager =
+        new DownloadManager(downloadCache, downloader, httpDownloader, eventHandler);
+    byte[] data = "content".getBytes(UTF_8);
+    doAnswer(
+            (Answer<Void>)
+                invocationOnMock -> {
+                  Path output = invocationOnMock.getArgument(5, Path.class);
+                  try (OutputStream outputStream = output.getOutputStream()) {
+                    ByteStreams.copy(new ByteArrayInputStream(data), outputStream);
+                  }
+                  return null;
+                })
+        .when(downloader)
+        .download(any(), any(), any(), any(), any(), any(), any(), any(), any(), eq("testRepo"));
+
+    Path result =
+        download(
+            downloadManager,
+            ImmutableList.of(URI.create("file:../vendored_lodash-4.17.21.tgz")),
+            ImmutableMap.of(),
+            ImmutableMap.of(),
+            Optional.empty(),
+            "testCanonicalId",
+            Optional.empty(),
+            fs.getPath(workingDir.newFile().getAbsolutePath()),
+            ImmutableMap.of(),
+            "testRepo");
+
+    assertThat(new String(ByteStreams.toByteArray(result.getInputStream()), UTF_8))
+        .isEqualTo("content");
+  }
+
+  @Test
+  public void downloadFromOpaqueFileUrl_withType() throws Exception {
+    Downloader downloader = mock(Downloader.class);
+    DownloadManager downloadManager =
+        new DownloadManager(downloadCache, downloader, httpDownloader, eventHandler);
+    byte[] data = "content".getBytes(UTF_8);
+    doAnswer(
+            (Answer<Void>)
+                invocationOnMock -> {
+                  Path output = invocationOnMock.getArgument(5, Path.class);
+                  try (OutputStream outputStream = output.getOutputStream()) {
+                    ByteStreams.copy(new ByteArrayInputStream(data), outputStream);
+                  }
+                  return null;
+                })
+        .when(downloader)
+        .download(any(), any(), any(), any(), any(), any(), any(), any(), any(), eq("testRepo"));
+
+    Path result =
+        download(
+            downloadManager,
+            ImmutableList.of(URI.create("file:../vendored_lodash-4.17.21.tgz")),
+            ImmutableMap.of(),
+            ImmutableMap.of(),
+            Optional.empty(),
+            "testCanonicalId",
+            Optional.of("tgz"),
+            fs.getPath(workingDir.newFolder().getAbsolutePath()),
+            ImmutableMap.of(),
+            "testRepo");
+
+    assertThat(result.getBaseName()).isEqualTo("vendored_lodash-4.17.21.tgz");
+    assertThat(new String(ByteStreams.toByteArray(result.getInputStream()), UTF_8))
+        .isEqualTo("content");
+  }
+
+  @Test
+  public void downloadFromOpaqueFileUrl_withEscapedQuestionMarkAndType() throws Exception {
+    Downloader downloader = mock(Downloader.class);
+    DownloadManager downloadManager =
+        new DownloadManager(downloadCache, downloader, httpDownloader, eventHandler);
+    byte[] data = "content".getBytes(UTF_8);
+    doAnswer(
+            (Answer<Void>)
+                invocationOnMock -> {
+                  Path output = invocationOnMock.getArgument(5, Path.class);
+                  try (OutputStream outputStream = output.getOutputStream()) {
+                    ByteStreams.copy(new ByteArrayInputStream(data), outputStream);
+                  }
+                  return null;
+                })
+        .when(downloader)
+        .download(any(), any(), any(), any(), any(), any(), any(), any(), any(), eq("testRepo"));
+
+    Path result =
+        download(
+            downloadManager,
+            ImmutableList.of(URI.create("file:../foo%3Fbar.tgz")),
+            ImmutableMap.of(),
+            ImmutableMap.of(),
+            Optional.empty(),
+            "testCanonicalId",
+            Optional.of("tgz"),
+            fs.getPath(workingDir.newFolder().getAbsolutePath()),
+            ImmutableMap.of(),
+            "testRepo");
+
+    assertThat(result.getBaseName()).isEqualTo("foo_bar.tgz");
+    assertThat(new String(ByteStreams.toByteArray(result.getInputStream()), UTF_8))
+        .isEqualTo("content");
+  }
+
+  @Test
+  public void getCandidateFileNames_opaqueFileUrlWithEscapedQuestionMark() {
+    assertThat(
+            DownloadManager.getCandidateFileNames(
+                URI.create("file:../foo%3Fbar.tgz"), fs.getPath("/tmp/foo_bar.tgz")))
+        .containsExactly("foo?bar.tgz", "foo_bar.tgz");
+  }
+
+  @Test
   public void downloadFrom2UrlsFirstOk() throws IOException, InterruptedException {
     try (ServerSocket server1 = new ServerSocket(0, 1, InetAddress.getByName(null));
         ServerSocket server2 = new ServerSocket(0, 1, InetAddress.getByName(null))) {

